### ISSUE TITLE
Specify port

### ISSUE THREAD

**Issue by Anuskuss** [marcussacana/DirectPackageInstaller#66]
Allow the user to specify the port instead of choosing a random one, e.g.
`dotnet DirectPackageInstaller.Desktop.dll -Server 192.168.0.1 -PS4 192.168.0.2 -Port 9090 http://192.168.0.1/ps4.pkg`

There's also the hardcoded port [9898](https://github.com/marcussacana/DirectPackageInstaller/blob/711bacfcfe812280414fbe36f4a6193354c40848/DirectPackageInstaller/DirectPackageInstaller/Host/PS4Server.cs#L38). Could you rewrite DPI such that all communication happens on a single port? Or if not, could you also make this configurable (e.g. `-Port 9898 -PayloadPort 9090`).

**Comment by marcussacana** [comment]
The process of DPI starts a download is a bit confusing as you can see:
```mermaid
flowchart TD
    B(Http Server) --> A[/DirectPackageInstaller/]   
    C(PKG Info Server) --> A
    D{HTTP PORT 9898} --> B
    E{Random Port} --> C 
    G(Payload Server 9090) --> F[GoldHEN]
    D --> J[/PS4 BGFT Download/]
    A --> G
    H[DPI PS4 Service] --> E
    H --> J
    F --> H
    B ~~~~ D
```
One of the key reason for that port be random is this complexity,
I wanted made the things simple as possible for the users does not need setup the envoriment,
Theorically the users shoudn't even lose his time setting the PS4/PC IP, even more with the ports.
The HTTP Port is fixed only due the fact that would broke the downloads from an older instance of DPI if not.
My preference is let the operating system handle that has possible and pick an port with enought privileges to listen it. 
There are another reason that made me favor the random port: Security
The DPI payload is basically an code that is injected in your system and install something automatically, the fact that the port is select randomly means that you doesn't expect to that "PKG Info Server" port be allowed in the router firewall, making the payload usable only behind LAN.
While is possible to me adapt to run in custom ports is really something that I have my worries, I need to think a bit about it.

**Comment by Anuskuss** [comment]
> the fact that the port is select randomly means that you doesn't expect to that "PKG Info Server" port be allowed in the router firewall, making the payload usable only behind LAN.

All my ports are closed (even for my LAN) unless I manually opened them. So I need to know which port is going to be used so I can open it.

> The HTTP Port is fixed only due the fact that would broke the downloads from an older instance of DPI if not.

More importantely, you are already listening on port TCP/9898. So you could theoretically halt the Payload server and let the HTTP server listen (or vice versa). Or just embed a minimal HTTP server (e.g. [civetweb](https://github.com/civetweb/civetweb)). Or even better give us the ability to specify `-JSON http://192.168.0.1/ps4.json -JSONFile /var/www/html/ps4.json` so you don't have to spin up a HTTP server at all. What do you think?

**Comment by marcussacana** [comment]
The PS4 payload can't start the download only with the JSON because it depends of some info that the DPI automatically take from the PKG, that's the job of that "PKG Info Server" in the flowchart.
So even if the DPI support send the json manifest directly, the download will not be able to start without the "PKG Info Server" job.
I don't think is the best way stop the HTTP server to run the PKG Info Server because it will interrupt the download for the user that tries to send a second download.
If you firewall is in your system and not the router I would hint you to just let the DPI open any port, the tool is open source and if is a trust problem is possible just keep a fork.
That said I will consider option to change the ports, but not sure of what to do.
I would like more let change in the Settings.ini instead cmd line if it is the case as well.

**Comment by Anuskuss** [comment]
> I don't think is the best way stop the HTTP server to run the PKG Info Server because it will interrupt the download for the user that tries to send a second download.

Isn't that only for situations where the PKG is served from DPI though? I send a link to my own HTTP server hosting the PKG so I wouldn't need that. I just find it ugly to open 2 ports for a single application (reminds me of FTP's ugly design) but if you can't make it work I'll deal with it.

> If you firewall is in your system and not the router I would hint you to just let the DPI open any port

I don't think that possible, at least not with `iptables`.

> if is a trust problem is possible just keep a fork.

It's not that I don't trust DPI, I don't trust any application period. But I can make an exception for DPI, and open 1 (or 2) port(s) that just DPI is allowed to use. But I gotta know which port to open so a `-Port` argument would be needed.

> I would like more let change in the Settings.ini instead cmd line

I don't like the idea of `Settings.ini`; I'll only ever need `-PS4`, `-Port` and `<url>`. I don't even want to specify `-Server` and rather let DPI figure it out itself (through the PS4 argument, e.g. `ip route show to match 192.168.0.2`). I'll call DPI from my PS4 through PHP and `Settings.ini` would make that needlessly complicated (I would have to edit `Settings.ini` from PHP instead of just calling the executable).

**Comment by marcussacana** [comment]
I will analyze the possibility.

**Comment by marcussacana** [comment]
The feature that you requested will be avaiable from the version 8.1.0 and later.
Currently the PayloadPort will be an hidden setting that you need include on Settings.ini file manually
With DPI Closed, open your `settings.ini` file and include something like this
`PayloadPort=9191`
Then DPI should stop using random ports for the connection.

**Comment by Anuskuss** [comment]
I don't like the idea of having it in `Settings.ini` but whatever.

So anyway seems like the latest version targets `.NET 8.0.0` now which I don't seem to be able to run:
```
$ dotnet DirectPackageInstaller.Desktop.dll -Server 192.168.0.1 -PS4 192.168.0.2 http://192.168.0.1/test.pkg
Unhandled exception. System.Exception: XOpenDisplay failed
   at Avalonia.X11.AvaloniaX11Platform.Initialize(X11PlatformOptions options)
   at Avalonia.AvaloniaX11PlatformExtensions.<>c.<UseX11>b__0_0()
   at Avalonia.AppBuilder.SetupUnsafe()
   at Avalonia.AppBuilder.Setup()
   at Avalonia.AppBuilder.SetupWithLifetime(IApplicationLifetime lifetime)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, Action`1 lifetimeBuilder)
   at DirectPackageInstaller.Desktop.Program.Main(String[] args) in /home/runner/work/DirectPackageInstaller/DirectPackageInstaller/DirectPackageInstaller/DirectPackageInstaller.Desktop/Program.cs:line 49
Aborted
```

The offending line
https://github.com/marcussacana/DirectPackageInstaller/blob/133c2772b4ede86cf1aed67b1ecbc8ca78c1698d/DirectPackageInstaller/DirectPackageInstaller.Desktop/Program.cs#L49
runs because of
https://github.com/marcussacana/DirectPackageInstaller/blob/133c2772b4ede86cf1aed67b1ecbc8ca78c1698d/DirectPackageInstaller/DirectPackageInstaller.Desktop/Program.cs#L36
which means that the application is somehow not getting my arguments. I don't have any other .NET application so I couldn't tell you if that's an issue with DPI specifically.

**Comment by marcussacana** [comment]
Some issue with linux huh, wait a bit, I will need have a time to look this.
By the way are you running in a Raspberry or some similar device or a normal linux desktop?

**Comment by Anuskuss** [comment]
It's a Debian server.

**Comment by marcussacana** [comment]
x11 or wayland?

**Comment by Anuskuss** [comment]
It's a server, meaning headless. I just need the CLI.

**Comment by marcussacana** [comment]
So the problem here it may be the DPI is trying to start the AvaloniaUI when there are no interface backend.
I see.

**Comment by Anuskuss** [comment]
Yes, DPI starts the GUI when it's started without arguments. But I am passing some so the problem is how you're retrieving them, which is probably related to the upgrade to `.NET 8.0.0`.

**Comment by marcussacana** [comment]
It should be fixed with the new version.
Also, I tried add the -DPIPort param, but not tested.
You can re-open the issue if there are any bugs

**Comment by Anuskuss** [comment]
I tried `-Port 10000 -DPIPort 10001` which didn't work. 

Interestingly enough if I do `-Port 9090` (which I've always used) together with the new `-DPIPort 9898` I get a crash instead:
```
DirectPacakgeInstaller v8.2.0 - CLI
LAN: 192.168.0.1
PS4: 192.168.0.2
Port: 9090
Source: http://192.168.0.1/test.pkg
Pushing: Test
Title ID: CUSA00000
Content ID: UP0000-CUSA00000_00-TEST
Type: Game Digital Application (gd)
Unhandled exception. System.Net.Sockets.SocketException (98): Address already in use
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at DirectPackageInstaller.Host.PayloadService.EnsureServer(String PCIP, String PS4IP) in /home/runner/work/DirectPackageInstaller/DirectPackageInstaller/DirectPackageInstaller/DirectPackageInstaller/Host/PayloadService.cs:line 221
   at DirectPackageInstaller.Host.PayloadService.SendPKGPayload(String PS4IP, String PCIP, String URL, Boolean Silent, Boolean AutoSplit) in /home/runner/work/DirectPackageInstaller/DirectPackageInstaller/DirectPackageInstaller/DirectPackageInstaller/Host/PayloadService.cs:line 38
   at DirectPackageInstaller.Desktop.Program.Main(String[] args) in /home/runner/work/DirectPackageInstaller/DirectPackageInstaller/DirectPackageInstaller/DirectPackageInstaller.Desktop/Program.cs:line 293
Aborted
```

**Comment by marcussacana** [comment]
Port 9898 is reserved for the DPI HTTP Server, you will need select other.
To works the port at -DPIPort should have incoming TCP permission.
If still not works should be just a bug, as said, I didn't tested.
But certain strangely, this error means the -DPIPort param is really working, shound't cause a trouble to set to other port.

**Comment by Anuskuss** [comment]
> Port 9898 is reserved for the DPI HTTP Server, you will need select other.

I don't understand this. I though that's what `-DPIPort` did? I see it's [still hardcoded](https://github.com/marcussacana/DirectPackageInstaller/blob/11f1089bedb1d432cfe3e078e93b21aa57aad6ef/DirectPackageInstaller/DirectPackageInstaller/Host/PS4Server.cs#L40), so what does it do? 

> To works the port at -DPIPort should have incoming TCP permission.

I just opened port 10000 and 10001, so 9090 and 9898 were closed. I didn't test it with no firewall because I assumed those 2 ports were all that were needed.

**Comment by marcussacana** [comment]
You need to keep 9898 open, The -DPIPort is not about the http server, but instead the DPI PKG Info

![Image](https://github.com/user-attachments/assets/c5ebba8f-de0a-4b68-80a0-0e8d920f8dfb)

**Comment by marcussacana** [comment]
As said, the PKG installation that DPI does is a bit complex,
There are a lot of steps until the download starts, you may wanna use your own http server but does not means you can just ignore the DPI embedded server, it still needed for the PS4 collect some metadata.
The DPI Http Server send metadata about the package split manifest
The DPI Pkg Info Server send metadata about the package content
The Port param is just about the Goldhen payload port.
Basically 3 ports needs to be used in your PC, one for outcoming, two for incoming.

**Comment by marcussacana** [comment]
If you really wanna skip the DPI Http Server you can start the DPI download with valid pkg manifest json files instead the pkg file.
if threre aren't any bug (since is a pretty old and unused feature) if you give as download link a pkg manifest the DPI will detect and skip the DPI Http Server step, and PS4 will automatically known that needs connect to your custom server to get the PKG, but still depends of the PKG Info server, that's unskippable step.

**Comment by Anuskuss** [comment]
If I do just `-Port 9090` (without `-DPIPort`) and open 9090 and 9898 in my firewall it still doesn't work. I don't really care what the ports are; they just need to be static so I can open them. I've been [hardcoding the port](https://github.com/marcussacana/DirectPackageInstaller/commit/d5d3d99a62d3cb7a2c9340f60b2f78485829d075) in my own personal build and I think I'll just keep on using that. I rarely play PS4 games anyway and I don't need any new features. I think it's best we move on and leave this issue closed. You should remove `-DPIPort` though as that seems rather pointless.

**Comment by marcussacana** [comment]
`-Port` is the goldhen payload port, from the PC, this is a TCP outcoming port.
You should set `-Port` paramter only if you aren't using goldhen payload server default port.

I said there are two incoming port and one outcoming
The -Port is the outcoming port for the goldhen payload server
The -DPIPort is a incoming port for the pkg info metadata
The 9898 is a fixed incoming port for the http server
